File: backend/app/utils/rich_text.py
from __future__ import annotations

from typing import Any


def extract_plain_text(node: Any) -> str:
    if node is None:
        return ""
    if isinstance(node, str):
        return node
    if isinstance(node, list):
        return " ".join(part for item in node for part in [extract_plain_text(item)] if part).strip()
    if not isinstance(node, dict):
        return ""

    pieces: list[str] = []
    if node.get("type") == "text":
        text = str(node.get("text", "")).strip()
        if text:
            pieces.append(text)
    for child in node.get("content", []) or []:
        text = extract_plain_text(child)
        if text:
            pieces.append(text)
    return " ".join(piece for piece in pieces if piece).strip()

File: frontend/src/utils/richText.ts
export function extractPlainText(node: unknown): string {
  return extractPlainTextLines(node).join("\n").replace(/\n{3,}/g, "\n\n").trim();
}

function extractPlainTextLines(node: unknown): string[] {
  if (node == null) {
    return [];
  }
  if (typeof node === "string") {
    return node.trim() ? [node.trim()] : [];
  }
  if (Array.isArray(node)) {
    return node
      .flatMap((item) => extractPlainTextLines(item))
      .filter(Boolean);
  }
  if (typeof node !== "object") {
    return [];
  }

  const record = node as { type?: string; text?: string; content?: unknown[] };
  if (record.type === "text" && typeof record.text === "string" && record.text.trim()) {
    return [record.text.trim()];
  }

  const childLines = (record.content ?? []).flatMap((child) => extractPlainTextLines(child));
  if (["paragraph", "heading", "listItem", "blockquote"].includes(record.type ?? "")) {
    const line = childLines.join(" ").replace(/\s+/g, " ").trim();
    return line ? [line] : [];
  }
  return childLines;
}
